From 271aeae28b09552b5fa0538da10d943620d88106 Mon Sep 17 00:00:00 2001
From: Kristian Rietveld <kris@lanedo.com>
Date: Sun, 17 Feb 2013 13:06:59 +0100
Subject: [PATCH 2/3] Never intervene in the event stream for legacy mice

This is necessary for devices (e.g. Mighty Mouse) which do emit
precise deltas but no phase.
---
 gdk/gdkwindow.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/gdk/gdkwindow.c b/gdk/gdkwindow.c
index 6483ea8..0f22246 100644
--- a/gdk/gdkwindow.c
+++ b/gdk/gdkwindow.c
@@ -10843,12 +10843,20 @@ proxy_button_event (GdkEvent *source_event,
    */
   if (type == GDK_SCROLL && source_event->scroll.has_deltas)
     {
+      gboolean legacy_mouse =
+          source_event->scroll.phase == GDK_EVENT_SCROLL_PHASE_NONE &&
+          source_event->scroll.momentum_phase == GDK_EVENT_SCROLL_PHASE_NONE;
+
       if (source_event->scroll.phase == GDK_EVENT_SCROLL_PHASE_START)
         {
           set_last_scroll_event_windows (display, pointer_window, event_win);
         }
-      else
+      else if (!legacy_mouse)
         {
+          /* Never override pointer and event windows for legacy devices
+           * which are not capable of momentum scrolling. (The windows
+           * will be NULL, because they have never been set).
+           */
           pointer_window = g_object_get_qdata (G_OBJECT (display),
                                                quark_last_scroll_pointer_window);
           event_win = g_object_get_qdata (G_OBJECT (display),
-- 
1.7.4.4

